README.md
# rooverlay
[![npm version](https://badge.fury.io/js/rooverlay.svg)](https://www.npmjs.com/package/rooverlay) [![Build Status](https://travis-ci.org/Tickaroo/rooverlay.svg?branch=master)](https://travis-ci.org/Tickaroo/rooverlay) [![codecov.io](https://codecov.io/github/Tickaroo/rooverlay/coverage.svg?branch=master)](https://codecov.io/github/Tickaroo/rooverlay?branch=master)

Overlay library that displays image/video galleries, html or iframes.

#### Dependencies

This library has no dependencies and still weights just ~15 KB unminified.

#### Compatibility

Recent browsers such as: IE 8+, Safari, Firefox & Chrome.


## Install

```bash
$ npm install --save rooverlay
```


## Usage

Below is a example of usage.

#### Image Gallery

```javascript
var Rooverlay = require('rooverlay');

var rooverlay = new Rooverlay({
  loop: true,
  slides: [{
    type: 'image',
    description: 'Lorem ipsum dolor sit amet, consetetur sadipscing elitr',
    src: 'rooverlay-1.jpg'
  }, {
    type: 'iframe-video',
    title: 'Roo',
    description: 'Lorem ipsum dolor sit amet, consetetur sadipscing elitr',
    src: 'https://d1zucocqrua2yq.cloudfront.net/v2/video/player-frame/50b6675694a940db6d000001/media-00ism10unt4h9ek0ysbr25?autoplay=true'
  }, {
    type: 'image',
    src: 'rooverlay-3.jpg'
  }]
});
```

#### Login Overlay

```javascript
var Rooverlay = require('rooverlay');

var rooverlay = new Rooverlay({
  loop: true,
  slides: [{
    type: 'iframe',
    src: 'https://staging-api.tickaroo.com/oauth/login?response_type=code&client_id=55d34d46e4b0b5f93ed111da&_lang=en'
  }]
});
```

## API

#### `new Rooverlay(options)`:

- **`options`**:
  - **`closeOnOverlayClick`**: Closes slideshow on overlay background click. (default `false`)
  - **`slideIndex`**: Initial start index of the slideshow slides. (default `0`)
  - **`loop`**: Loops slideshow. (default `false`)
  - **`skin`**: Slideshow layout skin, possible values are `'light'` and `'dark'` (default `'dark'`, `undefined`)
  - **`extraClasses`**: Adds extra classes to the class attribute of the root element `.rooverlay-wrapper` (default `undefined`)
  - **`pagination`**: Shows pagination. (default `false`)
  - **`paginationDescending`**: Counts the pages in descending order. (default `false`)
  - **`container`**: Element slideshow will be appended to. (default `document.body`)
  - **`disableKeyboardControls`**: Disable keyboard arrow key controls to go forward or back and the "escape" key to close the overlay. (default `false`)
  - **`onBeforeClose`**: Callback after closing the overlay. (default `undefined`)
  - **`onAfterSlideRender`**: Callback after slide renders. (default `undefined`)
  - **`onBeforeSlideRender`**: Callback before slide renders. (default `undefined`)
  - **`i18n`**:
    Text displayed to the user.
    (type `PlainObject`)
    - **`missing`**: Appears when content can't be loaded. (default `'Missing'`)
  - **`slides`**:
    Slides that will be displayed in the overlay.
    (type `Array<PlainObject>` default `[]`)  

  #### Image slide:
    - **`type`**: `'image'`
    - **`src`**: Image url. (default `undefined`)
    - **`width`**: Maximal image width. (default dynamic image width)
    - **`height`**: Maximal image height. (default dynamic image height)
    - **`minWidth`**: Minimal image width, if window size allows it. (default `300`)
    - **`minHeight`**: Minimal image height, if window size allows it. (default `300`)
    - **`aspectRatio`**: Whether or not to respect image size ratio on resize. (default `true`)
    - **`description`**: Description text/HTML string appended to the content element. (default `undefined`)
    - **`descriptionElement`**: Description DOM element appended to the content element. (default `undefined`)
    - **`title`**: Title text/HTML string appended to the top row title element. (default `undefined`)

  #### Video iframe slide:
    - **`type`**: `'iframe-video'`
    - **`src`**: Iframe url. (default `undefined`)
    - **`width`**: Maximal iframe width. (default `800`)
    - **`height`**: Maximal iframe height. (default `450`)
    - **`minWidth`**: Minimal iframe width, if window size allows it. (default `undefined`)
    - **`minHeight`**: Minimal iframe height, if window size allows it. (default `undefined`)
    - **`aspectRatio`**: Whether or not to respect iframe size ratio on resize. (default `true`)
    - **`description`**: Same as in `image`
    - **`descriptionElement`**: Same as in `image`
    - **`title`**: Same as in `image`

  #### Iframe slide:
    - **`type`**: `'iframe'`
    - **`src`**: Iframe url. (default `undefined`)
    - **`width`**: Maximal iframe width. (default `800`)
    - **`height`**: Maximal iframe height. (default `450`)
    - **`minWidth`**: Minimal iframe width, if window size allows it. (default `undefined`)
    - **`minHeight`**: Minimal iframe height, if window size allows it. (default `undefined`)
    - **`aspectRatio`**: Whether or not to respect iframe size ratio on resize. (default `false`)

  #### HTML slide:
    - **`type`**: `'html'`
    - **`html`**: HTML string that will be set inside the content element. (default `undefined`)
    - **`content`**: If `'html'` isn't set: DOM element that will be inserted to the content element. (default `undefined`)
    - **`width`**: Maximal HTML width. (default `600`)
    - **`height`**: Maximal HTML height. (default `undefined`)
    - **`minWidth`**: Minimal html width, if window size allows it. (default `undefined`)
    - **`minHeight`**: Minimal html height, if window size allows it. (default `undefined`)
    - **`aspectRatio`**: Whether or not to respect html size ratio on resize. (default `false`)


#### `instance.updateOptions(options)`:

Update the initial options.

- **`options`**: Same as in initialization.


#### `instance.updateSlides(slides)`:

Update the initial slides .
- **`slides`**: Same as in initialization `options.slides`.

#### `instance.updateSlidesAndRerenderWithIndex(slides, index)`:

Update the initial slides and rerenders them with slide index if `index` is set.

- **`slides`**: Same as in initialization `options.slides`.
- **`index`**: Slide index (default `undefined`)

#### `instance.nextSlide()`:

Go to the next slide.

#### `instance.previousSlide()`:

Go to the previous slide.

#### `instance.jumpToSlide(index)`:

Jump to the index.

- **`index`**: Slide index (default `0`)

#### `instance.fit()`:

Will center content element inside viewport.

#### `instance.showLoading()`:

Show loading spinner.

#### `instance.hideLoading()`:

Hides loading spinner.

#### `instance.destroy()` or `instance.close()`:

Close the overlay, remove it from the DOM and remove the event listeners, like `resize`.


## CSS

### Links

```sass
@import "~rooverlay/src/index.scss"
@import "~rooverlay/src/images.scss"
```

### Classes

- **`.rooverlay-overlay`**: Overlay background-color
- **`.rooverlay-description`**: Description box color, background-color and padding


HISTORY.md
## 1.2.3 / 2017-04-24

- Fix: use 'clientHeight' instead of 'scrollHeight' to avoid autocomplete jumping

## 1.2.2 / 2017-03-20

- Optimized `nextSlide` and `previousSlide` rendering
- Fixed slide html `type` content reset

## 1.2.1 / 2017-02-14

- Fixed `closeOnOverlayClick` option sometimes not working properly after `updateOptions()` was called

## 1.2.0 / 2017-02-09

- Added new option for `image` and `iframe-video` slides `descriptionElement`
- Added support for scrollable `html` and `iframe` slide types
- Locked page background scrolling

## 1.1.0 / 2016-12-30

- Added new option `extraClasses`
- Improved spinner layout (only works in browsers with css animation support)

## 1.0.0 / 2016-10-18

- Update to a major version to have better workflow with SemVer

## 0.2.6 / 2016-09-30

- IE8 fixes

## 0.2.5 / 2016-09-20

- Fixed IE `onload` callbacks unset
- Fixed internal click event handling

## 0.2.4 / 2016-09-20

- Internal click event handling changed

## 0.2.3 / 2016-09-20

- `z-index: 8000` added to `.rooverlay-wrapper`

## 0.2.2 / 2016-09-19

- `skin` option rendering improved
- Size calculation paddings improved

## 0.2.1 / 2016-09-15

- Size calculations for dimension without aspected ratio improved
- `paginationDescending` option added

## 0.2.0 / 2016-09-15

- Styling improved for `.rooverlay-description`
- "Escape" key support added
- **Breaking**: `updateOptions` `rerender` parameter removed

## 0.1.0 / 2016-09-14

- Pagination counter improved
- `skin` option added
- `slideIndex` option added
- `disableKeyboardControls` option added
- `minWidth` and `minHeight` options added to slides options

## 0.0.1 / 2016-09-06

- Init


src/index.js
function noop(){}
function debounce(duration, callback){
  clearTimeout(this.timerDebounce);
  this.timerDebounce = setTimeout(callback, duration);
}

function preventAndStopDefaultEvent(e){
  if (e.preventDefault){
    e.preventDefault();
  }
  if (e.stopPropagation){
    e.stopPropagation();
  }
}

function addEvent(element, eventName, callback){
  if (element.addEventListener){
    element.addEventListener(eventName, callback);
  }
  else {
    element.attachEvent('on' + eventName, callback);
  }
}

function removeEvent(element, eventName, callback){
  if (element.removeEventListener){
    element.removeEventListener(eventName, callback);
  }
  else {
    element.detachEvent('on' + eventName, callback);
  }
}

function Rooverlay(options){
  options = options || {};
  var wrapperElem = this.wrapperTemplate(this.getWrapperSkinClassName(options.skin, options.extraClasses));
  this.elems = {
    wrapper: wrapperElem,
    pagination: wrapperElem.querySelector('.rooverlay-pagination'),
    content: wrapperElem.querySelector('.rooverlay-content'),
    overlay: wrapperElem.querySelector('.rooverlay-overlay'),
    loader: wrapperElem.querySelector('.rooverlay-loader'),
    title: wrapperElem.querySelector('.rooverlay-title'),
    close: wrapperElem.querySelector('.rooverlay-close'),
    right: wrapperElem.querySelector('.rooverlay-right'),
    left: wrapperElem.querySelector('.rooverlay-left'),
    top: wrapperElem.querySelector('.rooverlay-top'),
    bottom: wrapperElem.querySelector('.rooverlay-bottom')
  };

  this.updateOptions(options);
  this.bindEvents();

  var self = this;
  this.resizeCallback = function(){
    debounce(100, function(){
      self.fit();
    });
  };
  this.keydownCallback = function(e){
    if (e.which === 27){
      e.preventDefault();
      self.destroy();
    }
    if (e.which === 37 || e.which === 38){
      e.preventDefault();
      self.previousSlide();
    }
    if (e.which === 39 || e.which === 40){
      e.preventDefault();
      self.nextSlide();
    }
  };
  addEvent(window, 'resize', this.resizeCallback);
  if ( ! this.options.disableKeyboardControls) {
    addEvent(window, 'keydown', this.keydownCallback);
  }

  document.documentElement.className = document.documentElement.className + ' rooverlay-enabled';

  this.options.container.appendChild(this.elems.wrapper);
  self.fit();
}

Rooverlay.prototype.getCurrentSlide = function getCurrentSlide(){
  return this.options.slides && this.options.slides[this.currentSlideIndex];
};

Rooverlay.prototype.updateSlides = function updateSlides(slides){
  if (slides && slides.length){
    this.lastSlideIndex = slides.length - 1;
    this.options.slides = slides;
  } else {
    this.currentSlideIndex = 0;
    this.lastSlideIndex = 0;
    this.options.slides = [];
  }
};

Rooverlay.prototype.updateSlidesAndRerenderWithIndex = function updateSlidesAndRerender(slides, i){
  this.updateSlides(slides);
  if (this.options.slides.length && i !== undefined){
    this.jumpToSlide(i);
  }
};

Rooverlay.prototype.updateOptions = function updateOptions(options){
  this.options = {
    closeOnOverlayClick: options.closeOnOverlayClick,
    skin:       options.skin,
    pagination: options.pagination,
    container:  options.container  || document.body,
    loop:       options.loop,
    i18n:       options.i18n       || {missing: 'Missing'},
    extraClasses: options.extraClasses || '',
    paginationDescending:    options.paginationDescending === true,
    disableKeyboardControls: options.disableKeyboardControls === true,
    onBeforeClose:       options.onBeforeClose       || noop,
    onAfterSlideRender:  options.onAfterSlideRender  || noop,
    onBeforeSlideRender: options.onBeforeSlideRender || noop
  };
  this.updateSlidesAndRerenderWithIndex(options.slides, options.slideIndex || 0);
};

Rooverlay.prototype.wrapperTemplate = function wrapperTemplate(classes){
  var elem = document.createElement('div');
  if (classes !== undefined) {
    elem.className = classes;
  }
  elem.innerHTML = '\
<div class="rooverlay-overlay"></div>\
<a href="#" class="rooverlay-left rooverlay-hide"></a>\
<a href="#" class="rooverlay-right rooverlay-hide"></a>\
<div class="rooverlay-top">\
  <div class="rooverlay-top-left">\
    <div class="rooverlay-pagination"></div>\
  </div>\
  <div class="rooverlay-top-center">\
    <div class="rooverlay-title"></div>\
  </div>\
  <div class="rooverlay-top-right">\
    <a href="#" class="rooverlay-close"></a>\
  </div>\
</div>\
<div class="rooverlay-bottom"></div>\
<div class="rooverlay-content"></div>\
<div class="rooverlay-loader rooverlay-hide"></div>';
  return elem;
};

Rooverlay.prototype.fit = function fit(){
  if ( ! this.slide) { return; }
  var holderWidth = window.innerWidth || document.documentElement.clientWidth || document.body.clientWidth;
  var holderHeight = window.innerHeight || document.documentElement.clientHeight || document.body.clientHeight;

  var sizeSettings = this._getSizeSettings(this.slide, this.imageElem);
  var marginHeight = this.getMarginHeight();

  if (this.slide.type === 'iframe' && holderHeight < 500) {
    marginHeight = 20;
  }

  var size = this._calculateSize(sizeSettings, holderWidth - this.getMarginWidth(), holderHeight - marginHeight);
  if ( ! size){
    return;
  }

  var topSpace = (holderHeight - size.height) / 2;
  if (this.slide.type === 'html' && holderHeight - marginHeight < sizeSettings.height) {
    if (this.elems.wrapper.className.indexOf('rooverlay-wrapper--scroll') === -1) {
      this.elems.wrapper.className = this.elems.wrapper.className + ' rooverlay-wrapper--scroll';
    }
    size.height = 0;
    topSpace = 10;
  }
  else {
    this.elems.wrapper.className = this.elems.wrapper.className.replace(new RegExp(' rooverlay-wrapper--scroll', 'g'), '');
  }

  this.elems.content.style.width = size.width + 'px';
  this.elems.content.style.height = size.height ? (size.height + 'px') : '';
  this.elems.content.style.minWidth = size.minWidth ? (size.minWidth + 'px') : '';
  this.elems.content.style.minHeight = size.minHeight ? (size.minHeight + 'px') : '';

  this.elems.content.style.top = topSpace ? (topSpace + 'px') : '';
  this.elems.content.style.left = (holderWidth - size.width) / 2 + 'px';
  this._balanceDescriptionHeight();
};

Rooverlay.prototype._balanceDescriptionHeight = function _balanceDescriptionHeight(){
  if (this.slide.description && this.slide.type === 'iframe-video') {
    var descriptionElem = this.elems.content.querySelector('.rooverlay-description');
    if (descriptionElem) {
      this.elems.content.style.marginTop =  ((descriptionElem.offsetHeight * -1) / 2) + 'px';
      return;
    }
  }
  this.elems.content.style.marginTop = '0px';
};

Rooverlay.prototype._getSizeSettings = function _getSizeSettings(slide, imageElem){
  var sizeSettings = {
    height: undefined,
    width: undefined,
    minWidth: slide.minWidth,
    minHeight: slide.minHeight,
    aspectRatio: false
  };
  switch (slide.type){
    case 'image':
      sizeSettings.width = slide.width || imageElem.naturalWidth || imageElem.width;
      sizeSettings.height = slide.height || imageElem.naturalHeight || imageElem.height;
      sizeSettings.minWidth = sizeSettings.minWidth || 300;
      sizeSettings.minHeight = sizeSettings.minHeight || 300;
      sizeSettings.aspectRatio = slide.aspectRatio || true;
    break;
    case 'iframe-video':
      sizeSettings.width = slide.width || 800;
      sizeSettings.height = slide.height || 450;
      sizeSettings.aspectRatio = slide.aspectRatio || true;
    break;
    case 'html':
      sizeSettings.width = slide.width || this.elems.content.lastChild.scrollWidth;
      if ( ! sizeSettings.width) {
        return sizeSettings;
      }
      sizeSettings.height = slide.height || this.elems.content.lastChild.clientHeight || sizeSettings.height;
      sizeSettings.aspectRatio = slide.aspectRatio || false;
    break;
    default:
      sizeSettings.width = slide.width || 600;
      sizeSettings.height = slide.height || sizeSettings.height;
      sizeSettings.aspectRatio = slide.aspectRatio || false;
    break;
  }
  return sizeSettings;
};

Rooverlay.prototype._calculateSize = function _calculateSize(sizeSettings, holderGlobalWidth, holderGlobalHeight){
  var height, minHeight, width, minWidth;
  var holderGlobalRatio = holderGlobalHeight / holderGlobalWidth;

  if (sizeSettings.aspectRatio && sizeSettings.height && sizeSettings.width){
    var ratio = (sizeSettings.height / sizeSettings.width);
    if (ratio > holderGlobalRatio){
      height = holderGlobalHeight;
      width = height / ratio;
    }
    else {
      width = holderGlobalWidth;
      height = width * ratio;
    }
    if (width > sizeSettings.width || height > sizeSettings.height){
      width = sizeSettings.width;
      height = sizeSettings.height;
    }
  }
  else {
    if (holderGlobalWidth > sizeSettings.width){
      width = sizeSettings.width;
    }
    else {
      width = holderGlobalWidth;
    }
    if (holderGlobalHeight > sizeSettings.height){
      height = sizeSettings.height;
    }
    else {
      height = holderGlobalHeight;
    }
  }

  if (width < sizeSettings.minWidth && sizeSettings.minWidth < holderGlobalWidth) {
    minWidth = sizeSettings.minWidth;
    width = sizeSettings.minWidth;
  }
  if (height < sizeSettings.minHeight && sizeSettings.minHeight < holderGlobalHeight) {
    minHeight = sizeSettings.minHeight;
    height = sizeSettings.minHeight;
  }

  return {
    width: width,
    minWidth: minWidth,
    height: height,
    minHeight: minHeight
  };
};

Rooverlay.prototype.checkArrows = function checkArrows(){
  var hasMany = this.options.slides.length > 1;
  if (this.options.loop && hasMany){
    this.elems.right.className = 'rooverlay-right';
    this.elems.left.className = 'rooverlay-left';
  }
  else if (hasMany){
    // right
    if (this.currentSlideIndex === this.lastSlideIndex){
      this.elems.right.className = 'rooverlay-right rooverlay-hide';
    }
    else {
      this.elems.right.className = 'rooverlay-right';
    }
    // left
    if (this.currentSlideIndex === 0){
      this.elems.left.className = 'rooverlay-left rooverlay-hide';
    }
    else {
      this.elems.left.className = 'rooverlay-left';
    }
  }
  else {
    this.elems.right.className = 'rooverlay-right rooverlay-hide';
    this.elems.left.className = 'rooverlay-left rooverlay-hide';
  }
};

Rooverlay.prototype.bindEvents = function bindEvents(){
  var self = this;
  addEvent(this.elems.left, 'click', function(e){
    preventAndStopDefaultEvent(e);
    self.previousSlide();
  });
  addEvent(this.elems.right, 'click', function(e){
    preventAndStopDefaultEvent(e);
    self.nextSlide();
  });
  addEvent(this.elems.close, 'click', function(e){
    preventAndStopDefaultEvent(e);
    self.destroy();
  });
  addEvent(this.elems.overlay, 'click', function(e){
    if (self.options.closeOnOverlayClick){
      preventAndStopDefaultEvent(e);
      self.destroy();
    }
  });
};

Rooverlay.prototype.previousSlide = function previousSlide(){
  this.currentSlideIndex--;
  if (this.options.loop){
    if (this.currentSlideIndex < 0){
      this.currentSlideIndex = this.lastSlideIndex;
    }
  }
  if (this.options.slides.length > 1) {
    this.renderSlide();
  }
};

Rooverlay.prototype.nextSlide = function nextSlide(){
  this.currentSlideIndex++;
  if ( ! this.options.loop){
    if (this.currentSlideIndex > this.lastSlideIndex){
      this.currentSlideIndex = this.lastSlideIndex;
    }
  }
  if (this.options.slides.length > 1) {
    this.renderSlide();
  }
};

Rooverlay.prototype.jumpToSlide = function jumpToSlide(index){
  this.currentSlideIndex = index || 0;
  this.renderSlide();
};

Rooverlay.prototype.showLoading = function showLoading(){
  this.elems.loader.className = 'rooverlay-loader';
};

Rooverlay.prototype.hideLoading = function hideLoading(){
  this.elems.loader.className = 'rooverlay-loader rooverlay-hide';
};

Rooverlay.prototype.showClose = function showClose(){
  this.elems.close.className = 'rooverlay-close';
};

Rooverlay.prototype.hideClose = function hideClose(){
  this.elems.close.className = 'rooverlay-close rooverlay-hide';
};

Rooverlay.prototype.getWrapperSkinClassName = function getWrapperSkinClassName(skinType, extraClasses){
  var classes = '';
  if (skinType === 'light') {
    classes = ' rooverlay-skin-light';
  }
  if (extraClasses) {
    classes += ' ' + extraClasses;
  }

  return 'rooverlay-wrapper' + classes;
};

Rooverlay.prototype.renderSlide = function renderSlide(){
  var slide = this.getCurrentSlide();
  if ( ! slide){
    if (this.currentSlideIndex !== 0){
      this.currentSlideIndex = 0;
      this.renderSlide();
    }
    return;
  }

  if (this.imageElem){
    this.imageElem.onload = function(){};
    this.imageElem.onerror = function(){};
  }
  if (this.iframeElem){
    this.iframeElem.onload = function(){};
    this.iframeElem.onerror = function(){};
  }

  var staticWrapperClasses = this.getWrapperSkinClassName(this.options.skin, this.options.extraClasses);
  this.elems.wrapper.className = staticWrapperClasses + ' rooverlay-type-' + slide.type;
  this.checkArrows();

  var self = this;
  this.slide = slide;
  this.options.onBeforeSlideRender();
  switch (slide.type){
    case 'image':
      this.showClose();
      this.showLoading();
      this.imageElem = new window.Image();
      this.imageElem.className = 'rooverlay-img';
      this.imageElem.onload = function(){
        self.elems.content.innerHTML = '';
        self.elems.content.appendChild(self.imageElem);
        self.hideLoading();
        self.appendDescription();
        self.afterSlideRender();
        self.fit();
      };
      this.imageElem.onerror = function(){
        self.elems.content.innerHTML = '<div class="rooverlay-missing">' + self.options.i18n.missing + '</div>';
        self.hideLoading();
        self.afterSlideRender();
        self.fit();
      };
      this.imageElem.src = slide.src;
    break;
    case 'iframe-video':
      self.elems.content.innerHTML = '';
      this.showClose();
      this.showLoading();
      this.updatePagination();
      this.iframeElem = document.createElement('iframe');
      this.iframeElem.className = 'rooverlay-video-frame';
      this.elems.content.appendChild(this.iframeElem);
      this.appendDescription();
      this.fit();
      this.iframeElem.src = slide.src;
      this.iframeElem.onload = function(){
        self.iframeElem.onload = function(){};
        self.hideLoading();
        self.afterSlideRender();
        self.fit();
      };
      this.iframeElem.onerror = function(){
        self.elems.content.innerHTML = '<div class="rooverlay-missing">' + self.options.i18n.missing + '</div>';
        self.hideLoading();
        self.afterSlideRender();
        self.fit();
      };
    break;
    case 'iframe':
      self.elems.content.innerHTML = '';
      this.showLoading();
      this.updatePagination();
      self.fit();
      this.iframeElem = document.createElement('iframe');
      this.iframeElem.onload = function(){
        self.iframeElem.onload = function(){};
        self.hideLoading();
        self.afterSlideRender();
        self.fit();
      };
      this.iframeElem.onerror = function(){
        self.elems.content.innerHTML = '<div class="rooverlay-missing">' + self.options.i18n.missing + '</div>';
        this.changeToContentClose();
        self.hideLoading();
        self.afterSlideRender();
        self.fit();
      };
      this.iframeElem.src = slide.src;
      self.elems.content.appendChild(self.iframeElem);
      this.changeToContentClose();
    break;
    case 'html':
      self.elems.content.innerHTML = '';
      var elem = document.createElement('div');
      elem.className = 'rooverlay-content-html';
      if (slide.html){
        elem.innerHTML = slide.html;
      }
      else if (slide.content){
        elem.appendChild(slide.content);
      }
      else {
        elem.innerHTML = '';
      }
      this.elems.content.appendChild(elem);
      this.changeToContentClose();
      this.hideLoading();
      this.afterSlideRender();
      self.fit();
    break;
  }
};

Rooverlay.prototype.changeToContentClose = function changeToContentClose(){
  var self = this;
  var contentCloseElem = document.createElement('a');
  contentCloseElem.href = '#';
  contentCloseElem.className = 'rooverlay-content-close';
  addEvent(contentCloseElem, 'click', function(e){
    preventAndStopDefaultEvent(e);
    self.destroy();
  });
  if (this.elems.content.firstChild){
    this.elems.content.insertBefore(contentCloseElem, this.elems.content.firstChild);
  }
  else {
    this.elems.content.appendChild(contentCloseElem);
  }
  this.hideClose();
};

Rooverlay.prototype.appendDescription = function appendDescription(){
  if (this.slide.description || this.slide.descriptionContent){
    var descriptionElem = document.createElement('div');
    descriptionElem.className = 'rooverlay-description';
    if (this.slide.descriptionContent) {
      descriptionElem.appendChild(this.slide.descriptionContent);
    }
    else {
      descriptionElem.innerHTML = this.slide.description;
    }
    this.elems.content.appendChild(descriptionElem);
  }
};

Rooverlay.prototype.afterSlideRender = function afterSlideRender(){
  if (this.slide.title){
    this.elems.title.innerHTML = this.slide.title;
  }
  else {
    this.elems.title.innerHTML = '';
  }
  this.updatePagination();
  this.options.onAfterSlideRender();
};

Rooverlay.prototype.updatePagination = function afterSlideRender(){
  if (this.options.pagination){
    if (this.options.paginationDescending){
      this.elems.pagination.innerHTML = (this.lastSlideIndex - this.currentSlideIndex + 1) + '/' + (this.lastSlideIndex + 1);
    }
    else {
      this.elems.pagination.innerHTML = (this.currentSlideIndex + 1) + '/' + (this.lastSlideIndex + 1);
    }
  }
};

Rooverlay.prototype.close = Rooverlay.prototype.destroy = function destroy(){
  this.options.onBeforeClose();
  this.options.container.removeChild(this.elems.wrapper);
  document.documentElement.className = document.documentElement.className.replace(' rooverlay-enabled', '');
  removeEvent(window, 'resize', this.resizeCallback);
  removeEvent(window, 'keydown', this.keydownCallback);
};

Rooverlay.prototype.getMarginWidth = function getMarginWidth(){
  if (this.options.marginWidth){
    return this.options.marginWidth;
  }
  if ( ! this.elems.right.offsetWidth){
    return (this.elems.left.offsetWidth || 10) * 2;
  }
  if ( ! this.elems.left.offsetWidth){
    return this.elems.right.offsetWidth * 2;
  }
  return this.elems.left.offsetWidth + this.elems.right.offsetWidth;
};

Rooverlay.prototype.getMarginHeight = function getMarginHeight(){
  if (this.options.marginHeight){
    return this.options.marginHeight;
  }
  if ( ! this.elems.bottom.offsetHeight){
    return (this.elems.top.offsetHeight || 10) * 2;
  }
  if ( ! this.elems.top.offsetHeight){
    return this.elems.bottom.offsetHeight * 2;
  }
  return this.elems.top.offsetHeight + this.elems.bottom.offsetHeight;
};

module.exports = Rooverlay;
